src/core/src/function.cpp

```diff
@@ -913,3 +913,112 @@ ov::Output<ov::Node> ov::Function::add_output(const ov::Output<ov::Node>& port)
+            merged_indexes.push_back(i);
+            // Not all dimensions can be merged
+            std::stringstream stream;
+            stream << "Get original batch size fails due to conflicting batch values for inputs:" << std::endl;
+            for (size_t j = 0; j < merged_indexes.size(); ++j) {
+                bs_util::dump_parameter(stream, f, merged_indexes[j]);
+            }
+            stream << "---" << std::endl;
+            stream << "Please ensure that N(Batch) dimension is set correctly for listed parameters";
+            OPENVINO_ASSERT(false, stream.str());
+        } else {
+            merged_indexes.push_back(i);
+        }
+        batch_initialized = true;
+    }
+    if (!batch_initialized) {
+        // Create graceful message to set layout for some parameters
+        std::stringstream stream;
+        stream << "Get original batch size fails due to batch is not set in any layout for any input. ";
+        stream << "Available inputs:" << std::endl;
+        for (size_t i = 0; i < f->get_parameters().size(); ++i) {
+            bs_util::dump_parameter(stream, f, i);
+        }
+        stream << "---" << std::endl;
+        stream << "Please use 'set_layout' API to set layout with batch dimension, e.g. "
+                  "`Function->get_parameters()[index]->set_layout(\"NCHW\");`";
+
+        OPENVINO_ASSERT(false, stream.str());
+    }
+    return batch_size;
+}
+
+void ov::util::set_batch_size(const std::shared_ptr<ov::Function>& f, ov::Dimension batch_size) {
+    get_batch_size(f);  // Ensure that function's batch size is valid and can be changed
+    // Now batch size can be set for all needed parameters
+    for (const auto& param : f->get_parameters()) {
+        const auto& layout = param->get_layout();
+        if (layout.empty() || !ov::layout::has_batch(layout))
+            continue;
+        const auto& pshape = param->get_partial_shape();
+        if (pshape.rank().is_dynamic()) {
+            continue;  // Parameter with fully dynamic rank can be left as is
+        }
+        auto batch_idx = bs_util::get_batch(layout, pshape);
+        param->get_partial_shape()[batch_idx] = batch_size;
+    }
+    try {
+        f->validate_nodes_and_infer_types();
+    } catch (const std::exception& e) {
+        std::stringstream stream;
+        stream << "Validation fails after set_batch_size. Possible reasons are:" << std::endl;
+        stream << "    1) Ensure that all inputs have valid layout set with batch dimension" << std::endl;
+        stream << "    2) Check model's documentation if batch size can be set to it at all" << std::endl;
+        stream << "Available inputs after set_batch_size:" << std::endl;
+        for (size_t i = 0; i < f->get_parameters().size(); ++i) {
+            bs_util::dump_parameter(stream, f, i);
+        }
+        stream << "---" << std::endl;
+        stream << "Original error message is: " << e.what();
+        OPENVINO_ASSERT(false, stream.str());
```

Comment:
I suppose we need to restore original shapes like we do in reshape? See https://github.com/openvinotoolkit/openvino/blob/master/src/core/src/function.cpp#L865-L867

BTW, why this method is not implemented via reshape? It calls smart reshape as well

Reply:
Good suggestion, added reusing of 'reshape'. Also enhanced error message to show original shape and intended shape after setting batch size